src/api/nbt_template.py

```diff
@@ -0,0 +1,333 @@
+
+    @property
+    def structure(self) -> Union[NBTStructure, Dict[str, NBTStructure]]:
+        return self._structure
+
+    @property
+    def default_value(self) -> Any:
+        return self._default_value
+
+    def apply_to(self, other_tag: NBTCompoundEntry) -> NBTCompoundEntry:
+        if (
+            isinstance(self.structure, NBTStructure)
+            and self.structure.tag_type == "any"
+        ):
+            return other_tag
+
+        for tag_name in self.structure.keys():
+            template_tag = self.structure[tag_name]
+            nbt_tag = other_tag.get(tag_name)
+
+            if nbt_tag:
+                if template_tag.tag_type == "any":
+                    continue
+
+                other_tag[tag_name] = template_tag.apply_to(nbt_tag)
+            elif template_tag.default_value is not None:
+                other_tag[tag_name] = template_tag.create_entry_from_template()
+
+        return other_tag
+
+    def create_entry_from_template(self, initial_value=None) -> NBTCompoundEntry:
+        if initial_value is None:
+            return NBTCompoundEntry(self.default_value)
+        return NBTCompoundEntry(initial_value)
+
+
+class NBTEntry:
+    """
+    Generic container used to represent a NBT entry that can have a template applied to it
+    """
+
+    __slots__ = ("tag_type", "value")
+
+    def __init__(self, tag_type: str, value: Any):
+        self.tag_type = tag_type
+        self.value = value
+
+    def __eq__(self, other: NBTEntry) -> bool:
+        return self.tag_type == other.tag_type and self.value == other.value
+
+    def __ne__(self, other: NBTEntry) -> bool:
+        return not self == other
+
+
+class NBTListEntry(UserList):
+    """
+    Container to represent a NBT List entry
+    """
+
+    tag_type = "list"
+
+    @property
+    def value(self):
+        return self.data
+
+
+class NBTCompoundEntry(UserDict):
+    """
+    Container to represent a NBT Compound entry
+    """
+
+    tag_type = "compound"
+
+    @property
+    def value(self):
+        return self.data
+
+
+class TemplateLoader:
+    """
+    Handles the loading of NBT template json files
+    """
+
+    def __init__(self, template_dir: str):
+        self._template_dir = template_dir
+
+        iter_chain = chain(
+            iglob(join(self._template_dir, "entities", "*.json")),
+            iglob(join(self._template_dir, "blockentities", "*.json")),
+            iglob(join(self._template_dir, "misc", "*.json")),
+        )
+        self._templates = {basename(p)[:-5]: p for p in iter_chain}
+        self._entity_template = self.load_template("entity")
+        self._item_template = self.load_template("item")
+
+    @property
+    def template_dir(self) -> str:
+        return self._template_dir
+
+    @property
+    def templates(self) -> Dict[str, str]:
+        return self._templates
+
+    def _tag_hook(self, json_object: dict) -> Union[dict, NBTStruct]:
+        if "tag_type" not in json_object:
```

Comment:
`_tag_hook()` is called on the deepest JSON object/array and then recurses back up and continually passed each new object to the function. The `if "tag_type" not in json_object` is just to prevent the json loader from trying to convert the `metdata` JSON object to a NBTStructure. That's also the only reason (and case) either dict or a `NBTStruct` is returned.

I decided to implement "metadata"/"extends" so when creating "creeper_nbt.json", there wasn't any need to copy over all of the previously defined/shared tags from another entity every time a new entity template was needed.

Reply:
I think I fully understand your question now, the reason `tag_type` is in the head of the file is because the `Explosion` tag in `item_nbt.json` has the tag_type `<explosion>` so that tag type is replaced by the contents (plus the head defined `tag_type`) of `explosion_nbt.json`. When a `tag_type` is in the form `<...>`, it replaces the entire tag data/value with the contents of the file referenced in the original `tag_type`